### ISSUE TITLE
labeling in GROMACS conversion

### ISSUE THREAD

**Issue by bannanc** [MobleyLab/SolvationToolkit#36]
I was trying to use solvation toolkit to make a box with water, cyclohexane, DMSO, and acetonitrile. 

Below is the script I used:

```
solvs = ['cyclohexane', 'water', 'DMSO', 'Acetonitrile']
smile_solv = ['C1CCCCC1', 'O', 'CS(=O)C', 'CC#N']
num_solv = [130, 780, 4, 2]

builder = MixtureSystem(solvs, smile_solv, num_solv, './allSolvents/', None)
builder.run(just_build = True)
builder.convert_to_gromacs()
```

For some reason in the GROMACS topology file, it labeled the DMSO 'water', the Acetonitrile 'water2', and the water 'water3', but then also had a molecules named 'DMSO' and 'Acetonitrile' which were more waters. I went back to look at the amber files. The prmtop file, while harder to read as a human, I think it looks fine. 

So I could get this simulation running, I followed the script in convert_to_gromacs method, but by hand with my labels and used parmed to convert the prmtop and inpcrd files to .top and .gro files and that topology file looks fine. I used the following lines to do this parmed conversion:

```
import parmed
gromacs_topology = parmed.load_file('tleap/cyclohexane_water_DMSO_Acetonitrile_130_780_4_2.prmtop', 'tleap/cyclohexane_water_DMSO_Acetonitrile_130_780_4_2.inpcrd')

components = gromacs_topology.split()

resnames = ['cyclohexane'] * 130
resnames += ['DMSO']*4
resnames += ['Acetonitrile'] * 2
resnames += ['water'] * 780
for i in range(len(resnames)):
     gromacs_topology.residues[i].name = resnames[i]

gromacs_topology.save('mol.top', format = 'gromacs')
gromacs_topology.save('mol.gro')
```

I have all of the files [here](https://www.dropbox.com/sh/4fkim2s558c2dz0/AADUb8aCX73Tz5vjjpP8nxhZa?dl=0)

Where the fromSolvationToolkit* are the ones created from the first script using a MixtureSystem
and the fromParmEd* are the two out put by my code above

I think part of the problem is that the prmtop files put the water at the end even though that wasn't the order they were input into the MixtureSystem, so the convert_to_gromacs tried to assign the label 'water' to the DMSO and Acetonitrile molecules since they showed up before the water and then assigned DMSO and Acetonitrile to the last few waters.

**Comment by bannanc** [comment]
@davidlmobley

**Comment by davidlmobley** [comment]
Thanks. I'll look at this, though it may take me a couple of days. I
probably made an assumption that molecules weren't being re-ordered by
tleap, which sounds like it's not true.

Thanks.


On Fri, Apr 1, 2016 at 1:48 PM, Caitlin Bannan <notifications@github.com>
wrote:

> I was trying to use solvation toolkit to make a box with water,
> cyclohexane, DMSO, and acetonitrile.
>
> Below is the script I used:
>
> solvs = ['cyclohexane', 'water', 'DMSO', 'Acetonitrile']
> smile_solv = ['C1CCCCC1', 'O', 'CS(=O)C', 'CC#N']
> num_solv = [130, 780, 4, 2]
>
> builder = MixtureSystem(solvs, smile_solv, num_solv, './allSolvents/', None)
> builder.run(just_build = True)
> builder.convert_to_gromacs()
>
> For some reason in the GROMACS topology file, it labeled the DMSO 'water',
> the Acetonitrile 'water2', and the water 'water3', but then also had a
> molecules named 'DMSO' and 'Acetonitrile' which were more waters. I went
> back to look at the amber files. The prmtop file, while harder to read as a
> human, I think it looks fine.
>
> So I could get this simulation running, I followed the script in
> convert_to_gromacs method, but by hand with my labels and used parmed to
> convert the prmtop and inpcrd files to .top and .gro files and that
> topology file looks fine. I used the following lines to do this parmed
> conversion:
>
> import parmed
> gromacs_topology = parmed.load_file('tleap/cyclohexane_water_DMSO_Acetonitrile_130_780_4_2.prmtop', 'tleap/cyclohexane_water_DMSO_Acetonitrile_130_780_4_2.inpcrd')
>
> components = gromacs_topology.split()
>
> resnames = ['cyclohexane'] * 130
> resnames += ['DMSO']*4
> resnames += ['Acetonitrile'] * 2
> resnames += ['water'] * 780
> for i in range(len(resnames)):
>      gromacs_topology.residues[i].name = resnames[i]
>
> gromacs_topology.save('mol.top', format = 'gromacs')
> gromacs_topology.save('mol.gro')
>
> I have all of the files here
> <https://www.dropbox.com/sh/4fkim2s558c2dz0/AADUb8aCX73Tz5vjjpP8nxhZa?dl=0>
>
> Where the fromSolvationToolkit* are the ones created from the first script
> using a MixtureSystem
> and the fromParmEd* are the two out put by my code above
>
> I think part of the problem is that the prmtop files put the water at the
> end even though that wasn't the order they were input into the
> MixtureSystem, so the convert_to_gromacs tried to assign the label 'water'
> to the DMSO and Acetonitrile molecules since they showed up before the
> water and then assigned DMSO and Acetonitrile to the last few waters.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/MobleyLab/SolvationToolkit/issues/36>
>



-- 
David Mobley
Associate Professor
Department of Pharmaceutical Sciences
Department of Chemistry
3134B Natural Sciences I
University of California, Irvine
Irvine, CA 92697
dmobley@uci.edu
work (949) 824-6383
cell (949) 385-2436

**Comment by davidlmobley** [comment]
I've verified that after ParmEd read of the AMBER files, the water appears at the end of the list of components. And I've verified that this happens due to `tleap`, which apparently recognizes water and moves it to the end (special treatment of water strikes again). The next step is for me to figure out a robust way to catch this and then handle components separately. I'm working on it.

**Comment by davidlmobley** [comment]
OK, I think I know how to fix this now. `tleap` apparently reorders any residue with residue name `WAT` to move to the end of the file, which includes all water models it recognizes as such (i.e. TIP3P and SPC, etc.). So the converter can look for this residue name and recognize a re-ordering has taken place. Potentially this might also have to deal with flexible GAFF water or other water NOT receiving special treatment by `tleap`; i.e. if someone invoked `openmoltools.amber.build_mixture_prmtop` and asked for the water not to receive special treatment. But in that case, it will come through with some other residue name (i.e. "water", unless it is "WAT" in which case `tleap` will re-order it). So I think we should be able to recognize this way.
Next I'll try and fix it.

**Comment by davidlmobley** [comment]
Resolved via #47 .
